File: src/main/java/fi/metatavu/metaform/server/metaforms/FieldRuleEvaluator.java
package fi.metatavu.metaform.server.metaforms;

import java.util.List;
import java.util.Map;
import java.util.Objects;

import fi.metatavu.metaform.api.spec.model.FieldRule;
import fi.metatavu.metaform.api.spec.model.Reply;
import org.apache.commons.lang3.StringUtils;


/**
 * Field rule evaluator
 * 
 * @author Antti Leppä
 */
public class FieldRuleEvaluator {

  /**
   * Evaluates field rule
   * 
   * @param rule  rule
   * @param reply reply data
   * @return evaluation result
   */
  public boolean evaluate(FieldRule rule, Reply reply) {
    Map<String, Object> data = reply.getData();
    String field = rule.getField();
    Object fieldData = data.get(field);

    boolean result = false;
    String fieldValue = Objects.toString(fieldData, null);
    String ruleEquals = rule.getEquals();
    String ruleNotEquals = rule.getNotEquals();

    List<FieldRule> andRules = rule.getAnd();
    List<FieldRule> orRules = rule.getOr();

    if (StringUtils.isNotBlank(field) && StringUtils.isNotBlank(ruleEquals)) {
      result = StringUtils.equals(ruleEquals, fieldValue);
    }

    if (StringUtils.isNotBlank(field) && StringUtils.isNotBlank(ruleNotEquals)) {
      result = !StringUtils.equals(ruleNotEquals, fieldValue);
    }

    if (andRules != null) {
      for (int i = 0; i < andRules.size(); i++) {
        if (!evaluate(andRules.get(i), reply)) {
          return false;
        }
      }
    }

    if (orRules != null) {
      for (int i = 0; i < orRules.size(); i++) {
        if (evaluate(orRules.get(i), reply)) {
          return true;
        }
      }
    }

    return result;
  }

}

File: src/main/kotlin/fi/metatavu/metaform/server/metaform/FieldRuleEvaluator.kt
package fi.metatavu.metaform.server.metaform

import fi.metatavu.metaform.api.spec.model.FieldRule
import fi.metatavu.metaform.api.spec.model.Reply
import org.apache.commons.lang3.StringUtils
import java.util.*

/**
 * Field rule evaluator
 *
 * @author Antti Leppä
 */
class FieldRuleEvaluator {
    /**
     * Evaluates field rule
     *
     * @param rule rule
     * @param reply reply data
     * @return evaluation result
     */
    fun evaluate(rule: FieldRule, reply: Reply): Boolean {
        val data = reply.data
        val field = rule.field
        val fieldData = data?.get(field)
        var result = false
        val fieldValue = Objects.toString(fieldData, null)
        val ruleEquals = rule.equals
        val ruleNotEquals = rule.notEquals
        if (StringUtils.isNotBlank(field) && StringUtils.isNotBlank(ruleEquals)) {
            result = StringUtils.equals(ruleEquals, fieldValue)
        }
        if (StringUtils.isNotBlank(field) && StringUtils.isNotBlank(ruleNotEquals)) {
            result = !StringUtils.equals(ruleNotEquals, fieldValue)
        }
        rule.and?.forEach { andRule ->
            if (!evaluate(andRule, reply)) {
                return false
            }
        }

        rule.or?.forEach { orRule ->
            if (evaluate(orRule, reply)) {
                return true
            }
        }
        return result
    }
}
